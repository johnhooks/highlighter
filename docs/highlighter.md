<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@bitmachina/highlighter](./highlighter.md)

## highlighter package

## Functions

|  Function | Description |
|  --- | --- |
|  [createHighlighter(options)](./highlighter.createhighlighter.md) | Create MDsveX code highlighting function. |
|  [parseMetadata(metastring)](./highlighter.parsemetadata.md) | Parse Markdown code fence metadata. |
|  [renderToHast({ tokens, options, metadata, })](./highlighter.rendertohast.md) | Render Shiki tokens to hast, Hypertext Abstract Syntax Tree format. |

## Interfaces

|  Interface | Description |
|  --- | --- |
|  [HastOptions](./highlighter.hastoptions.md) | Options to modify the syntax tree of the hast code block. |
|  [HastRendererOptions](./highlighter.hastrendereroptions.md) |  |
|  [LineOption](./highlighter.lineoption.md) |  |

## Type Aliases

|  Type Alias | Description |
|  --- | --- |
|  [MdvexHighlighter](./highlighter.mdvexhighlighter.md) | <p>MDsveX highlighter type.</p><p>NOTE: not exported from <code>mdsvex</code>, so its copied here.</p> |
|  [Metadata](./highlighter.metadata.md) | Markdown Code Block Metadata |

