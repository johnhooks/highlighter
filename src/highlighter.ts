import type { Element } from "hast";
import { toHtml } from "hast-util-to-html";
import rangeParser from "parse-numeric-range";
import rehypeParse from "rehype-parse";
import { getHighlighter, type HighlighterOptions } from "shiki";
import { unified } from "unified";
import { visit } from "unist-util-visit";

/**
 * MDsveX highlighter type.
 *
 * NOTE: not exported from `mdsvex`, so its copied here.
 */
type Highlighter = (
  code: string,
  lang: string | undefined,
  metastring: string | undefined
) => string | Promise<string>;

/**
 * Escape curlies, backtick, \t, \r, \n to avoid breaking output of {@html `here`} in .svelte
 *
 * [reference](https://github.com/pngwn/MDsveX/blob/6c60fe68c335fce559db9690fbf5e69ef539d37d/packages/mdsvex/src/transformers/index.ts#L571)
 * @param str - The string to escape.
 */
function escape_svelty(str: string): string {
  return str
    .replace(
      /[{}`]/g,
      (c) => ({ "{": "&#123;", "}": "&#125;", "`": "&#96;" }[c as unknown as "{" | "}" | "`"])
    )
    .replace(/\\([trn])/g, "&#92;$1");
}

const reverseString = (s) => s?.split("").reverse().join("");

const hastParser = unified().use(rehypeParse, { fragment: true });

/**
 * Create MDsveX code highlighting function.
 *
 * @param options - Shiki highlighter options object.
 * @returns A Promise of a MDsvex highlighter function.
 * @public
 */
export async function createHighlighter(options: HighlighterOptions): Promise<Highlighter> {
  const shikiHighlighter = await getHighlighter(options);

  /**
   * Highlighter function for Mdsvex codeblocks.
   *
   * [reference](https://github.com/atomiks/rehype-pretty-code/blob/e24f8415c8264dc868006ced839d3cb7445e716a/src/index.js#L50)
   * @param raw - Source code to highlight
   * @param lang - Source language
   * @param metastring - Source metadata
   * @returns HTML string
   */
  return function highlighter(
    raw: string,
    lang: string | undefined,
    metastring: string | undefined
  ): string {
    const titleMatch = metastring?.match(/title="(.+)"/);
    const title = titleMatch?.[1] ?? null;
    const titleString = titleMatch?.[0] ?? "";
    const metaWithoutTitle = metastring?.replace(titleString, "");

    const tree = hastParser.parse(shikiHighlighter.codeToHtml(raw, { lang: lang || "txt" }));

    const lineNumbers = metastring ? rangeParser(metaWithoutTitle?.match(/{(.*)}/)?.[1] ?? "") : [];
    let lineCounter = 0;

    // inspiration: https://github.com/atomiks/rehype-pretty-code/blob/e24f8415c8264dc868006ced839d3cb7445e716a/src/index.js#L89
    visit(tree, "element", (node) => {
      node.properties = node.properties ?? {};

      if (
        node.tagName === "code" &&
        /srebmuNeniLwohs(?!(.*)(\/))/.test(reverseString(metastring))
      ) {
        node.properties["data-line-numbers"] = "";

        const lineNumbersStartAtMatch = reverseString(metastring).match(
          /(?:\}(\d+){)?srebmuNeniLwohs(?!(.*)(\/))/
        );
        if (lineNumbersStartAtMatch[1])
          node.properties["style"] = `counter-set: line ${
            reverseString(lineNumbersStartAtMatch[1]) - 1
          };`;
      }

      // TODO fix the highlighted line to fill the scrolling overflow.
      if (node.properties.className?.[0] === "line") {
        if (lineNumbers.length !== 0 && lineNumbers.includes(++lineCounter)) {
          node.properties["data-highlight"] = "";
        }
      }
    });

    toFragment({ node: tree as unknown as Element, lang, title });

    return escape_svelty(toHtml(tree));
  };
}

/**
 * Convert hast `Root` to an `Element`.
 *
 * [reference](https://github.com/atomiks/rehype-pretty-code/blob/e24f8415c8264dc868006ced839d3cb7445e716a/src/index.js#L9)
 * @param params - Parameters to turn `node` into a `fragment`.
 * @returns
 */
function toFragment({
  node,
  lang,
  title,
  inline = false,
}: {
  node: Element;
  lang: string | undefined;
  title: string | null;
  inline?: boolean;
}) {
  node.tagName = inline ? "span" : "div";
  // User can replace this with a real Fragment at runtime
  node.properties = { "data-code-fragment": "" };

  const pre = node.children[0] as Element;
  // Remove class="shiki" and the background-color
  pre.properties = pre.properties ?? {};
  pre.properties.class = undefined;
  pre.properties["data-language"] = lang;

  const code = pre.children[0] as Element;
  code.properties = code.properties ?? {};
  code.properties["data-language"] = lang;

  if (inline) {
    node.children = [code];
    return;
  }

  if (title) {
    // TODO create a flexible format to customize adding a title to the code blocks.
    pre.properties["data-code-title"] = title;
    return;
  }

  node.children = [pre];
}
