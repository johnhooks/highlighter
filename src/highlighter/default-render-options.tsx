import type { IRenderOptions } from "./types.js";

/**
 * The default renderers for {@link renderToHast}
 *
 * @internal
 */
export const defaultRenderOptions: IRenderOptions = {
  pre({ style, children, meta }) {
    return meta.title ? (
      <div>
        <pre style={style} dataCodeTitle={meta.title}>
          {children}
        </pre>
      </div>
    ) : (
      <pre style={style}>{children}</pre>
    );
  },

  code({ children, meta }) {
    return <code dataLanguage={meta.lang}>{children}</code>;
  },

  line({ children, meta, index }) {
    const { lineNumbers, lineNumbersStart } = meta;
    const lineNumber = index + lineNumbersStart;
    const highlighted = lineNumbers?.includes(index) ? true : null;
    const style = { "counter-set": lineNumber };

    return (
      <span style={style} dataHighlighted={highlighted} dataLineNumber={lineNumber}>
        {children}
      </span>
    );
  },

  token({ style, children }) {
    return <span style={style}>{children}</span>;
  },
};
