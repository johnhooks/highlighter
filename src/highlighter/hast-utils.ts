/**
 * A super simple, probably naive hast implementation.
 *
 * {@link https://github.com/syntax-tree/hast-util-to-html/tree/fa187494263aa1087ba1c307abce72b2413e452e#tohtmltree-options | toHtml} escapes `&` and `<` and doesn't allow
 * control of HTML entities.
 */

import { html, find } from "property-information";

import { escapeHtml } from "./utils.js";

export type TProperty = string | string[] | boolean | undefined;
export type TProperties = Record<string, TProperty>;
export type TText = { type: "text"; value: string };
export type TElement = {
  type: "element";
  tagName: string;
  properties: TProperties;
  children: Array<TNode>;
};
export type TNode = TElement | TText;
export type TChild = TNode | string;
export type TChildren = TChild[];

/**
 * Render {@link TText} node.
 */
function renderText({ value }: TText): string {
  return escapeHtml(value);
}

/**
 * Render HTML attributes.
 */
function renderAttributes(attributes: TProperties): string {
  const result: string[] = [];

  for (const [name, value] of Object.entries(attributes)) {
    if (value === undefined || value === false) continue;

    const { attribute, booleanish, commaSeparated } = find(html, name);

    if (value === true) {
      if (booleanish || /^data-/.test(attribute)) {
        result.push(attribute);
      }
    } else if (Array.isArray(value)) {
      const seperator = commaSeparated ? "," : " ";
      result.push(`${attribute}="${value.join(seperator)}"`);
    } else {
      result.push(`${attribute}="${value}"`);
    }
  }

  if (result.length === 0) return "";

  // Add a space to separate from the tag.
  return result.join(" ");
}

function renderNode(node: TNode) {
  if (node.type === "text") {
    return renderText(node);
  } else {
    // This is pretty hacky, but it works for now.
    const newline = "dataLineNumber" in node.properties ? "\n" : "";
    const attributes = renderAttributes(node.properties);
    const children = node.children.map(renderNode).join("");
    const opening = node.tagName + (attributes.length > 1 ? " " + attributes : "");
    return `<${opening}>${children}</${node.tagName}>${newline}`;
  }
}

export function toHtml(node: TElement) {
  return renderNode(node);
}

export function h(tagName: string, properties: TProperties, children?: TChildren): TElement {
  const nodes: Array<TNode> = children
    ? children.map((child) => (typeof child === "string" ? { type: "text", value: child } : child))
    : [];

  return {
    type: "element",
    tagName,
    properties,
    children: nodes,
  };
}
