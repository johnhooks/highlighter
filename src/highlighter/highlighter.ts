import { toHtml } from "hast-util-to-html";
import { getHighlighter, type HighlighterOptions } from "shiki";

import { parseMetadata } from "../meta-parser/index.js";

import { renderToHast } from "./render-to-hast.js";

/**
 * MDsveX highlighter type.
 *
 * NOTE: not exported from `mdsvex`, so its copied here.
 * @public
 */
export type MdvexHighlighter = (
  code: string,
  lang: string | undefined,
  metastring: string | undefined
) => string | Promise<string>;

/**
 * Create MDsveX code highlighting function.
 *
 * @param options - Shiki highlighter options object.
 * @returns A Promise of a MDsvex highlighter function.
 * @public
 */
export async function createHighlighter(options: HighlighterOptions): Promise<MdvexHighlighter> {
  const shikiHighlighter = await getHighlighter(options);

  /**
   * Highlighter function for Mdsvex codeblocks.
   *
   * [reference](https://github.com/atomiks/rehype-pretty-code/blob/e24f8415c8264dc868006ced839d3cb7445e716a/src/index.js#L50)
   * @param code - Raw source code to highlight.
   * @param lang - Programming language.
   * @param metastring - Code block fence metadata.
   * @returns HTML string.
   */
  return function highlighter(
    code: string,
    lang: string | undefined,
    metastring: string | undefined
  ): string {
    if (!lang) return `<pre><code>${code}</code></pre>`;

    const tokens = shikiHighlighter.codeToThemedTokens(code, lang || "txt");
    const metadata = parseMetadata(metastring);
    const tree = renderToHast({ tokens, metadata });

    return toHtml(tree);
  };
}
