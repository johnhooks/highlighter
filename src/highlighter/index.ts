export * from "./highlighter.js";
export * from "./render-to-hast.js";
export * from "./types.js";
