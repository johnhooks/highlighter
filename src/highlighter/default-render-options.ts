import { h } from "./hast-utils.js";
import { TRenderOptions } from "./types";

/**
 * The default renderers for {@link renderToHast}
 *
 * @internal
 */
export const defaultRenderOptions: TRenderOptions = {
  pre({ children, metadata: { title: dataCodeTitle } }) {
    return h("pre", { dataCodeTitle }, children);
  },

  code({ children, metadata: { lang: dataLanguage, showLineNumbers: dataLineNumbers } }) {
    return h("code", { dataLanguage, dataLineNumbers }, children);
  },

  line({ children, index, metadata: { lineNumbers, lineNumbersStart } }) {
    const dataLineNumber = String(index + lineNumbersStart);
    const dataHighlighted = lineNumbers?.includes(index);
    return h("span", { dataHighlighted, dataLineNumber }, children);
  },

  token({ style, children }) {
    return h("span", { style }, children);
  },
};
