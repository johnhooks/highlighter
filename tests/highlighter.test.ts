import { vi } from "vitest";

import { createHighlighter } from "../src/highlighter/highlighter.js";

const bitmachinaHighlighter = await createHighlighter({ theme: "css-variables" });

describe("highlighter", () => {
  const highlighter = vi.fn(bitmachinaHighlighter);

  describe("render options", async () => {
    describe("line numbers", async () => {
      beforeEach(async () => {
        const code = `let line = 1;\nline = 2;\nline = 3;`;
        const html = await highlighter(code, "js");
        document.body.innerHTML = html;
      });

      describe("default", async () => {
        it("should add data-line-number properties", () => {
          const lines = document.querySelectorAll("span[data-line-number]");
          expect(lines.length).toEqual(3);
        });

        it("should start at the right line number", () => {
          const line = document.querySelector("span[data-line-number]");
          expect(line).toBeVisible();
          expect(line.getAttribute("data-line-number")).toEqual("1");
        });
      });

      describe("metadata provided", async () => {
        beforeEach(async () => {
          const code = `let line = 1;\nline = 2;\nline = 3;`;
          const html = await highlighter(code, "js", "js showLineNumbers{10}");
          document.body.innerHTML = html;
        });

        it("should add data-line-number properties", () => {
          const lines = document.querySelectorAll("span[data-line-number]");
          expect(lines.length).toEqual(3);
        });

        it("should start at the right line number", () => {
          const line = document.querySelector("span[data-line-number]");
          expect(line).toBeVisible();
          expect(line.getAttribute("data-line-number")).toEqual("10");
        });
      });
    });
  });

  describe("highlighted lines", async () => {
    const code = `let line = 1;\nline = 2;\nline = 3;\nline=4\nline=5\nline=6`;

    it("should highlight a single line", async () => {
      const html = await highlighter(code, "js", "js highlight{2}");
      document.body.innerHTML = html;
      const lines = document.querySelectorAll("span[data-line-numbers]");
      lines.forEach((line, index) => {
        expect(line.getAttribute("data-highlighted")).toBe(index === 1);
      });
    });

    it("should highlight a range of lines", async () => {
      const html = await highlighter(code, "js", "js highlight{2-4}");
      document.body.innerHTML = html;
      const lines = document.querySelectorAll("span[data-line-numbers]");
      lines.forEach((line, index) => {
        expect(line.getAttribute("data-highlighted")).toBe([1, 2, 3].includes(index));
      });
    });

    it("should highlight a single line and a range of lines", async () => {
      const html = await highlighter(code, "js", "js highlight{2-4,6}");
      document.body.innerHTML = html;
      const lines = document.querySelectorAll("span[data-line-numbers]");
      lines.forEach((line, index) => {
        expect(line.getAttribute("data-highlighted")).toBe([1, 2, 3, 5].includes(index));
      });
    });
  });
});
