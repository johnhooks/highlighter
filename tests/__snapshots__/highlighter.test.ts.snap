// Vitest Snapshot v1

exports[`createHighlighter > should render fixture: fixture.c 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-token-keyword)\\">#include</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;lt;stdio.h&#x26;gt;</span></span><span style=\\"counter-set: 1\\"></span><span style=\\"counter-set: 2\\"><span style=\\"color: var(--shiki-token-keyword)\\">int</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">main</span><span style=\\"color: var(--shiki-color-text)\\">()</span></span><span style=\\"counter-set: 3\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#123;</span></span><span style=\\"counter-set: 4\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-function)\\">printf(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;Hello World&#x26;quot;</span><span style=\\"color: var(--shiki-token-function)\\">)</span><span style=\\"color: var(--shiki-color-text)\\">;</span></span><span style=\\"counter-set: 5\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-keyword)\\">return</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-constant)\\">0</span><span style=\\"color: var(--shiki-color-text)\\">;</span></span><span style=\\"counter-set: 6\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#125;</span></span><span style=\\"counter-set: 7\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.cpp 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-token-keyword)\\">#include</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;lt;iostream&#x26;gt;</span></span><span style=\\"counter-set: 1\\"></span><span style=\\"counter-set: 2\\"><span style=\\"color: var(--shiki-token-keyword)\\">int</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">main</span><span style=\\"color: var(--shiki-color-text)\\">()</span></span><span style=\\"counter-set: 3\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#123;</span></span><span style=\\"counter-set: 4\\"><span style=\\"color: var(--shiki-color-text)\\">  std</span><span style=\\"color: var(--shiki-token-punctuation)\\">::</span><span style=\\"color: var(--shiki-color-text)\\">cout </span><span style=\\"color: var(--shiki-token-keyword)\\">&#x26;lt;&#x26;lt;</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;Hello World&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">;</span></span><span style=\\"counter-set: 5\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-keyword)\\">return</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-constant)\\">0</span><span style=\\"color: var(--shiki-color-text)\\">;</span></span><span style=\\"counter-set: 6\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#125;</span></span><span style=\\"counter-set: 7\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.css 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-token-function)\\">.fixture::after</span><span style=\\"color: var(--shiki-color-text)\\"> &#x26;#123;</span></span><span style=\\"counter-set: 1\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-constant)\\">content</span><span style=\\"color: var(--shiki-token-keyword)\\">:</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;hello world&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">;</span></span><span style=\\"counter-set: 2\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#125;</span></span><span style=\\"counter-set: 3\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.html 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;lt;!</span><span style=\\"color: var(--shiki-token-string-expression)\\">DOCTYPE</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">html</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 1\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;lt;</span><span style=\\"color: var(--shiki-token-string-expression)\\">html</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 2\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;lt;</span><span style=\\"color: var(--shiki-token-string-expression)\\">head</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 3\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;lt;/</span><span style=\\"color: var(--shiki-token-string-expression)\\">head</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 4\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;lt;</span><span style=\\"color: var(--shiki-token-string-expression)\\">body</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 5\\"><span style=\\"color: var(--shiki-color-text)\\">    &#x26;lt;</span><span style=\\"color: var(--shiki-token-string-expression)\\">h1</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;Hello World&#x26;lt;</span><span style=\\"color: var(--shiki-token-string-expression)\\">h1</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 6\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;lt;/</span><span style=\\"color: var(--shiki-token-string-expression)\\">body</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 7\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;lt;/</span><span style=\\"color: var(--shiki-token-string-expression)\\">html</span><span style=\\"color: var(--shiki-color-text)\\">&#x26;gt;</span></span><span style=\\"counter-set: 8\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.js 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-token-keyword)\\">function</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">hello</span><span style=\\"color: var(--shiki-color-text)\\">(value) &#x26;#123;</span></span><span style=\\"counter-set: 1\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-keyword)\\">if</span><span style=\\"color: var(--shiki-color-text)\\"> (</span><span style=\\"color: var(--shiki-token-keyword)\\">typeof</span><span style=\\"color: var(--shiki-color-text)\\"> value </span><span style=\\"color: var(--shiki-token-keyword)\\">===</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;string&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">) &#x26;#123;</span></span><span style=\\"counter-set: 2\\"><span style=\\"color: var(--shiki-color-text)\\">    </span><span style=\\"color: var(--shiki-token-constant)\\">console</span><span style=\\"color: var(--shiki-token-function)\\">.log</span><span style=\\"color: var(--shiki-color-text)\\">(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#96;Hello </span><span style=\\"color: var(--shiki-token-keyword)\\">$&#x26;#123;</span><span style=\\"color: var(--shiki-color-text)\\">value</span><span style=\\"color: var(--shiki-token-keyword)\\">&#x26;#125;</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#96;</span><span style=\\"color: var(--shiki-color-text)\\">);</span></span><span style=\\"counter-set: 3\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;#125; </span><span style=\\"color: var(--shiki-token-keyword)\\">else</span><span style=\\"color: var(--shiki-color-text)\\"> &#x26;#123;</span></span><span style=\\"counter-set: 4\\"><span style=\\"color: var(--shiki-color-text)\\">    </span><span style=\\"color: var(--shiki-token-keyword)\\">throw</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-keyword)\\">new</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">Error</span><span style=\\"color: var(--shiki-color-text)\\">(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#96;Can not greet &#x26;#39;undefined&#x26;#39;!&#x26;#96;</span><span style=\\"color: var(--shiki-color-text)\\">);</span></span><span style=\\"counter-set: 5\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;#125;</span></span><span style=\\"counter-set: 6\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#125;</span></span><span style=\\"counter-set: 7\\"></span><span style=\\"counter-set: 8\\"><span style=\\"color: var(--shiki-token-function)\\">hello</span><span style=\\"color: var(--shiki-color-text)\\">(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;world!&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">);</span></span><span style=\\"counter-set: 9\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.rb 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-token-function)\\">puts</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#39;Hello World&#x26;#39;</span></span><span style=\\"counter-set: 1\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.sh 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-color-text)\\">file=</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;fixture&#x26;quot;</span></span><span style=\\"counter-set: 1\\"><span style=\\"color: var(--shiki-color-text)\\">extensions=(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;css&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;js&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;sh&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;sh&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">)</span></span><span style=\\"counter-set: 2\\"></span><span style=\\"counter-set: 3\\"><span style=\\"color: var(--shiki-token-keyword)\\">for</span><span style=\\"color: var(--shiki-color-text)\\"> ext </span><span style=\\"color: var(--shiki-token-keyword)\\">in</span><span style=\\"color: var(--shiki-color-text)\\"> $&#x26;#123;extensions[*]&#x26;#125; </span><span style=\\"color: var(--shiki-token-keyword)\\">;</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-keyword)\\">do</span></span><span style=\\"counter-set: 4\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-function)\\">echo</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;$file.$&#x26;#123;ext&#x26;#125;&#x26;quot;</span></span><span style=\\"counter-set: 5\\"><span style=\\"color: var(--shiki-token-keyword)\\">done</span></span><span style=\\"counter-set: 6\\"></span><span style=\\"counter-set: 7\\"></span></code></pre>"`;

exports[`createHighlighter > should render fixture: fixture.ts 1`] = `"<pre><code><span style=\\"counter-set: 0\\"><span style=\\"color: var(--shiki-token-keyword)\\">function</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">hello</span><span style=\\"color: var(--shiki-color-text)\\">(value</span><span style=\\"color: var(--shiki-token-keyword)\\">:</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-constant)\\">string</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-keyword)\\">|</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-constant)\\">undefined</span><span style=\\"color: var(--shiki-color-text)\\">) &#x26;#123;</span></span><span style=\\"counter-set: 1\\"><span style=\\"color: var(--shiki-color-text)\\">  </span><span style=\\"color: var(--shiki-token-keyword)\\">if</span><span style=\\"color: var(--shiki-color-text)\\"> (</span><span style=\\"color: var(--shiki-token-keyword)\\">typeof</span><span style=\\"color: var(--shiki-color-text)\\"> value </span><span style=\\"color: var(--shiki-token-keyword)\\">===</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;string&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">) &#x26;#123;</span></span><span style=\\"counter-set: 2\\"><span style=\\"color: var(--shiki-color-text)\\">    </span><span style=\\"color: var(--shiki-token-constant)\\">console</span><span style=\\"color: var(--shiki-token-function)\\">.log</span><span style=\\"color: var(--shiki-color-text)\\">(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#96;Hello </span><span style=\\"color: var(--shiki-token-keyword)\\">$&#x26;#123;</span><span style=\\"color: var(--shiki-color-text)\\">value</span><span style=\\"color: var(--shiki-token-keyword)\\">&#x26;#125;</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#96;</span><span style=\\"color: var(--shiki-color-text)\\">);</span></span><span style=\\"counter-set: 3\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;#125; </span><span style=\\"color: var(--shiki-token-keyword)\\">else</span><span style=\\"color: var(--shiki-color-text)\\"> &#x26;#123;</span></span><span style=\\"counter-set: 4\\"><span style=\\"color: var(--shiki-color-text)\\">    </span><span style=\\"color: var(--shiki-token-keyword)\\">throw</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-keyword)\\">new</span><span style=\\"color: var(--shiki-color-text)\\"> </span><span style=\\"color: var(--shiki-token-function)\\">Error</span><span style=\\"color: var(--shiki-color-text)\\">(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;#96;Can not greet &#x26;#39;undefined&#x26;#39;!&#x26;#96;</span><span style=\\"color: var(--shiki-color-text)\\">);</span></span><span style=\\"counter-set: 5\\"><span style=\\"color: var(--shiki-color-text)\\">  &#x26;#125;</span></span><span style=\\"counter-set: 6\\"><span style=\\"color: var(--shiki-color-text)\\">&#x26;#125;</span></span><span style=\\"counter-set: 7\\"></span><span style=\\"counter-set: 8\\"><span style=\\"color: var(--shiki-token-function)\\">hello</span><span style=\\"color: var(--shiki-color-text)\\">(</span><span style=\\"color: var(--shiki-token-string-expression)\\">&#x26;quot;world!&#x26;quot;</span><span style=\\"color: var(--shiki-color-text)\\">);</span></span><span style=\\"counter-set: 9\\"></span></code></pre>"`;
